### DISCUSSION TITLE
Trying to understand how to represent "catalogs" in pgstac

### DISCUSSION THREAD

**Discussion by tylere** [stac-utils/.github#3] [Q&A]
_(starting the discussion here, since the pgstac repo does not have discussions enabled...)_

I'm trying to understand how STAC Catalogs (and subcatalogs) are represented ing `pgstac`.

The STACspec defines [Catalog](https://github.com/radiantearth/stac-spec/blob/master/catalog-spec/catalog-spec.md)s, [Collection](https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md)s, and [Item](https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md)s. A STAC Catalog object represents a logical group of other Catalog, Collection, and Item objects.

`pgstac` allows for the storage of Collections ([table](https://github.com/stac-utils/pgstac/blob/9effb17436076f57e1a272d6cd5b272380bea3d1/src/pgstac/sql/002_collections.sql#L11)) and Items ([table](https://github.com/stac-utils/pgstac/blob/9effb17436076f57e1a272d6cd5b272380bea3d1/src/pgstac/sql/003a_items.sql#L1)) which [pypgstac can load data into](https://github.com/stac-utils/pgstac/blob/9effb17436076f57e1a272d6cd5b272380bea3d1/src/pypgstac/python/pypgstac/pypgstac.py#L63). However, the concept of a Catalog seems to be largely missing from `pgstac`, and is only mentioned indirectly in the pgstac docs:

> The [STAC FastAPI](https://github.com/stac-utils/stac-fastapi) PgSTAC backend and [Franklin](https://github.com/azavea/franklin) can be used to expose a PgSTAC catalog. (from [README.md](https://github.com/stac-utils/pgstac/blob/9effb17436076f57e1a272d6cd5b272380bea3d1/README.md?plain=1#L31))

> If two or more collections in your catalog share a property name... (from [docs/src/pgstac.md](https://github.com/stac-utils/pgstac/blob/9effb17436076f57e1a272d6cd5b272380bea3d1/docs/src/pgstac.md?plain=1#L126)](https://github.com/stac-utils/pgstac/blob/9effb17436076f57e1a272d6cd5b272380bea3d1/docs/src/pgstac.md?plain=1#L126)

Given that background, here are a few specific questions:

- Does a pgstac database correspond to a single STAC Catalog?
- How can STAC Catalogs that contain subcatalogs be represented in pgstac?
- How can STAC Catalogs that only contain Items (i.e. no Collections) be represented in pgstac?

**Comment by gadomski** [comment]
> Does a pgstac database correspond to a single STAC Catalog?

Generally, yes. **pgstac** was developed primarily to be a backend to a STAC API, and most STAC APIs are represent a single `Catalog`.

> How can STAC Catalogs that contain subcatalogs be represented in pgstac?

I'd recommend reading through https://github.com/stac-utils/stac-fastapi/discussions/595 to get some more context on the subcatalog problem space. Tl;dr no one is working on it as far as I know.

> How can STAC Catalogs that only contain Items (i.e. no Collections) be represented in pgstac?

I _think_ that most of the STAC core team would consider this an anti-pattern (citation needed).

**Comment by gadomski** [comment]
@tylere NB https://github.com/radiantearth/stac-spec/discussions/categories/stac-software where we've discussed general STAC software topics previously — I realize there's no link to that page on the **stac-utils** README, I'll fix.

**Comment by tylere** [comment]
> > How can STAC Catalogs that only contain Items (i.e. no Collections) be represented in pgstac?
> 
> I _think_ that most of the STAC core team would consider this an anti-pattern (citation needed).

I didn't see this mentioned on the [STAC Best Practices](https://github.com/radiantearth/stac-spec/blob/master/best-practices.md) page. Rather, the [Catalog Layout](https://github.com/radiantearth/stac-spec/blob/master/best-practices.md#catalog-layout) section explicitly mentions the use of sub-catalogs containing items.

**Reply by tylere** [reply]
Thanks, it may make sense to transfer this discussion over there (but I am going to hold off doing so until after the next STAC Community Meeting).

**Reply by gadomski** [reply]
> I didn't see this mentioned on the [STAC Best Practices](https://github.com/radiantearth/stac-spec/blob/master/best-practices.md) page. Rather, the [Catalog Layout](https://github.com/radiantearth/stac-spec/blob/master/best-practices.md#catalog-layout) section explicitly mentions the use of sub-catalogs containing items.

I think this is another case where the best practices for STAC and for STAC APIs are different. Because APIs are so dependent on their implementation (i.e. the underlying server and database software), there tends to be (in my experience) less flexibility when working with a STAC API.
